xmtp_mls/src/groups/sync.rs

```diff
@@ -825,29 +835,41 @@ impl MlsGroup {
 
                     return Err(err);
                 }
-                Ok(Some((payload, post_commit_data))) => {
-                    let payload_slice = payload.as_slice();
+                Ok(Some(PublishIntentData {
+                    message,
+                    post_commit_action,
+                    staged_commit,
+                })) => {
+                    let payload_slice = message.as_slice();
+                    let has_staged_commit = staged_commit.is_some();
+                    provider.conn_ref().set_group_intent_published(
+                        intent.id,
+                        sha256(payload_slice),
+                        post_commit_action,
+                        staged_commit,
+                        openmls_group.epoch().as_u64() as i64,
+                    )?;
+                    log::debug!(
+                        "client [{}] set stored intent [{}] to state `published`",
+                        client.inbox_id(),
+                        intent.id
+                    );
 
                     client
                         .api_client
                         .send_group_messages(vec![payload_slice])
                         .await?;
+
                     log::info!(
                         "[{}] published intent [{}] of type [{}]",
                         client.inbox_id(),
                         intent.id,
                         intent.kind
                     );
-                    provider.conn_ref().set_group_intent_published(
-                        intent.id,
-                        sha256(payload_slice),
-                        post_commit_data,
-                    )?;
-                    log::debug!(
-                        "client [{}] set stored intent [{}] to state `published`",
-                        client.inbox_id(),
-                        intent.id
-                    );
+                    if has_staged_commit {
+                        log::info!("Canceling all further publishes, since a commit was found");
+                        return Err(GroupError::PublishCancelled);
+                    }
```

Comment:
This works if this method (publish_intents) is only called once, but won't further publishes still happen if `publish_intents` is called again? Then we would skip past this published intent because it is no longer in `TO_PUBLISH` state?

Reply:
Also, are we sure we want to return an error here, rather than returning success? Won't an error here stop us from calling `receive()` within `sync_with_conn()`, and finishing off the intent lifecycle?

Reply:
We continue through publish_intents errors in `sync_with_conn`. In the old method we would get them frequently if you had two intents lined up that both created commits, because they would trigger `PendingCommit` errors.

I did actually make that change in one of the down-stack PRs just to clean up error handling.